receiver/receivercreator/discovery.go

```diff
@@ -0,0 +1,183 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package receivercreator // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/receivercreator"
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/go-viper/mapstructure/v2"
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v3"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+)
+
+const (
+	// hints prefix
+	otelHints = "io.opentelemetry.discovery"
+
+	// hint suffix for metrics
+	otelMetricsHints = otelHints + ".metrics"
+
+	// hints definitions
+	discoveryEnabledHint = "enabled"
+	scraperHint          = "scraper"
+	configHint           = "config"
+)
+
+// hintsTemplatesBuilder creates configuration templates from provided hints.
+type hintsTemplatesBuilder interface {
+	createReceiverTemplateFromHints(env observer.EndpointEnv) (*receiverTemplate, error)
+}
```

Comment:
This interface doesn't seem like being used anywhere, right?

Reply:
right now it's only implemented by the `k8sHintsBuilder`. The idea is to give the option for implementing it in the various different environments like k8s, docker etc. This feature would make sense for docker as well.

Reply:
But `k8sHintsBuilder` is not being used at all. Existing functions operate on the struct instead. I don't see why we need to keep it around. It can be added when needed